src/librbd/Lock.cc

```diff
@@ -0,0 +1,620 @@
+typename Lock<L>::Action Lock<L>::get_active_action() const {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+  return m_actions_contexts.front().first;
+}
+
+template <typename L>
+void Lock<L>::complete_active_action(State next_state, int r) {
+  assert(m_lock.is_locked());
+  assert(!m_actions_contexts.empty());
+
+  ActionContexts action_contexts(std::move(m_actions_contexts.front()));
+  m_actions_contexts.pop_front();
+  m_state = next_state;
+
+  m_lock.Unlock();
+  for (auto ctx : action_contexts.second) {
+    ctx->complete(r);
+  }
+  m_lock.Lock();
+
+  if (!is_transition_state() && !m_actions_contexts.empty()) {
+    execute_next_action();
+  }
+}
+
+template <typename L>
+bool Lock<L>::is_shutdown_locked() const {
+  assert(m_lock.is_locked());
+
+  return ((m_state == STATE_SHUTDOWN) ||
+          (!m_actions_contexts.empty() &&
+           m_actions_contexts.back().first == ACTION_SHUT_DOWN));
+}
+
+template <typename L>
+void Lock<L>::send_acquire_lock() {
+  assert(m_lock.is_locked());
+  if (m_state == STATE_LOCKED) {
+    complete_active_action(STATE_LOCKED, 0);
+    return;
+  }
+
+  ldout(m_cct, 10) << this << " " << __func__ << dendl;
+  m_state = STATE_ACQUIRING;
+
+  if (!m_watcher->is_registered()) {
+    ldout(m_cct, 10) << this << " watcher not registered - delaying request"
+                     << dendl;
+    m_state = STATE_WAITING_FOR_REGISTER;
+    m_watcher->register_watch(new FunctionContext([this](int r) {
+          if (r < 0) {
+            lderr(m_cct) << "watcher registering error: " << cpp_strerror(r)
+                         << dendl;
+
+            Mutex::Locker locker(m_lock);
+            complete_active_action(STATE_UNLOCKED, r);
+            return;
+          }
+          m_work_queue->queue(new FunctionContext([this](int r) {
+            Mutex::Locker locker(m_lock);
+            send_acquire_lock();
+          }));
+    }));
+    return;
+  }
+
+  m_cookie = m_watcher->encode_lock_cookie();
+  AcquireRequest<L>* req = AcquireRequest<L>::create(
+    m_ioctx, m_watcher, m_oid, m_cookie,
+    util::create_context_callback<Lock<L>, &Lock<L>::handle_acquire_lock>(this));
+  m_work_queue->queue(new C_SendRequest<AcquireRequest<L>>(req), 0);
+}
+
+template <typename L>
+void Lock<L>::handle_acquire_lock(int r) {
+  ldout(m_cct, 10) << this << " " << __func__ << ": r=" << r << dendl;
+
+  if (r == -EBUSY || r == -EAGAIN) {
+    ldout(m_cct, 5) << "unable to acquire exclusive lock" << dendl;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to acquire exclusive lock:" << cpp_strerror(r)
+               << dendl;
+  } else {
+    ldout(m_cct, 5) << "successfully acquired exclusive lock" << dendl;
+  }
+
+  {
+    m_lock.Lock();
+    assert(m_state == STATE_ACQUIRING);
+
+    Action action = get_active_action();
+    assert(action == ACTION_TRY_LOCK || action == ACTION_REQUEST_LOCK);
+    if (action == ACTION_REQUEST_LOCK && r < 0 && r != -EBLACKLISTED &&
+        r != -EPERM) {
```

Comment:
I think the ```-EPERM``` was an error that would only be returned by the journal policy, so that check should be moved up to ```ExclusiveLock```